### PR TITLE
Move name resolution into phase 2

### PR DESCRIPTION
**Pull request by jseyfried**
r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
This "unfixes" #33231, i.e. the following will error with `rustc -Z no-analysis`:
```rust
extern crate doesnt_exist; //~ ERROR can't find crate
fn main() {}
```

Currently, the following errors with `rustc -Z no-analysis`:
```rust
#[macro_use] extern crate doesnt_exist; //~ ERROR can't find crate
fn main() {}
```

I could keep #33231 fixed in this PR by checking for `no-analysis` in `phase_2_configure_and_expand`, but it would cause problems later on when expansion depends on import resolution, for example:
```rust
extern crate doesnt_exist; //< We need to know if this crate has a top-level module `bar` ...
mod bar {
    macro_rules! m { () => {} }
}
fn main() {
    use doesnt_exist::*; //< ... (which would be imported here) ...
    bar::m!(); //< ... before we can expand this.
}
```

File: src/librustc_driver/driver.rs [hunk 1]
```diff
@@ -95,15 +95,15 @@ pub fn compile_input(sess: &Session,
...
-            };
+        let krate = match phase_1_parse_input(sess, cfg, input) {
+            Ok(krate) => krate,
+            Err(mut parse_error) => {
+                parse_error.emit();
+                return Err(1);
+            }
+        };
 
+        let krate = {
```
Comment (LeoTestard) [inline src/librustc_driver/driver.rs]: Is there any particular reason for this change besides style?

File: src/libsyntax/ext/expand.rs [hunk 2]
```diff
@@ -1212,24 +1211,6 @@ fn mark_tts(tts: &[TokenTree], m: Mrk) -> Vec<TokenTree> {
...
-}
-
-impl<'a, 'v> Visitor<'v> for MacroExterminator<'a> {
-    fn visit_mac(&mut self, mac: &ast::Mac) {
-        self.sess.span_diagnostic.span_bug(mac.span,
-                                           "macro exterminator: expected AST \
-                                           with no macro invocations");
-    }
-}
-
```
Comment (LeoTestard) [inline src/libsyntax/ext/expand.rs]: Why did you remove this pass?

File: src/librustc_driver/driver.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_driver/driver.rs]: No.

File: src/libsyntax/ext/expand.rs (hunk 2 above)
Comment (jseyfried) [inline src/libsyntax/ext/expand.rs]: I don't think it's useful. If we accidentally leave a macro in the AST after expansion, the [folds](https://github.com/rust-lang/rust/blob/81aeb0cddea5a614a4eebc0f0384fa27a7eb011f/src/libsyntax/fold.rs#L173-L180) and [visits](https://github.com/rust-lang/rust/blob/81aeb0cddea5a614a4eebc0f0384fa27a7eb011f/src/libsyntax/visit.rs#L102-L109) already panic. `resolve` and `lowering` also panic appropriately on unexpanded macros.

File: src/librustc_driver/driver.rs [hunk 3]
```diff
@@ -574,19 +546,28 @@ fn count_nodes(krate: &ast::Crate) -> usize {
 // For continuing compilation after a parsed crate has been
 // modified
 
+pub struct Phase2Results<'a> {
```
Comment (nrc) [inline src/librustc_driver/driver.rs]: Could this have a more descriptive name? PostExpansionResults? or ExpansionResult or ExpandedCrate, etc.

**Comment by nrc** [comment]
r=me with the renamed struct.

File: src/librustc_driver/driver.rs (hunk 3 above)
Comment (jseyfried) [inline src/librustc_driver/driver.rs]: Sure, renamed to `ExpansionResult`.

**Comment by SiegeLord** [comment]
I really wish I was pinged about this. I spent a lot of effort fixing this issue, and now my projects don't compile because my fix was reverted.

**Comment by jseyfried** [comment]
@SiegeLord sorry you weren't pinged.

We will need to load all `extern crate`s (not just `#[macro_use]` crates) during macro expansion to implement [RFC 1561](https://github.com/rust-lang/rfcs/pull/1561). `rustc -Z no-analysis` does expansion, so it will need to load all `extern crate`s.

If you want to avoid loading `extern crate`s, you will have to use `rustc -Z parse-only`, which does not do expansion. However, `rustc -Z parse-only --emit dep-info` does not emit anything because macro expansion can introduce new non-inline modules (e.g. `mod foo;`), which affect the dep-info.

That being said, I believe it would be possible to have `rustc -Z parse-only --emit dep-info` emit the dep-info on a best effort basis. Would this work for your use case?

**Comment by SiegeLord** [comment]
I'm sure there's a solution, but that's not really the issue here. In my fix I specifically added a test so that this wouldn't get re-broken, but if there's an option to remove the test then there's literally 0 motivation to fix this again, because then somebody else will just remove the test next time it's convenient for them. I'll just stop using this broken feature.

**Comment by jseyfried** [comment]
@SiegeLord 
Your test wasn't removed just because it was convenient, it was removed because it is *not feasible* to support the feature that it was testing due to the upcoming improvements to macros.
As I described [here](https://github.com/rust-lang/rust/pull/33997#issuecomment-222858143), we will *need* to load all `extern crate`s, not just `#[macro_use]` `extern crate`s to be able to perform macro expansion.

That being said, as I stated earlier, you should have been pinged so we could discuss ways to still support your use case.

I am offering to implement `rustc -Z parse-only --emit dep-info` emitting the dep info on a best effort basis. This would be feasible to support indefinitely.
